src/cache/RrObjectCache.js

```diff
@@ -0,0 +1,44 @@
+import validateCacheSize from './util/validateCacheSize';
+import isStringOrNumber from './util/isStringOrNumber';
+
+export default class RrObjectCache {
+  constructor({cacheSize} = {}) {
+    validateCacheSize(cacheSize);
+    this.clear();
+    this._cacheSize = cacheSize;
+  }
+  set(key, selectorFn) {
+    if (this._cacheLength >= this._cacheSize) {
+      this._randomReplace(key, selectorFn);
+    } else {
+      this._cache[key] = selectorFn;
+      this._cacheKeys[this._cacheLength] = key;
+      this._cacheLength++;
+    }
+  }
+  get(key) {
+    return this._cache[key];
+  }
+  remove(key) {
+    const index = this._cacheKeys.indexOf(key); // O(1)
+    if (index > -1) {
+      delete this._cache[key];
+      this._cacheLength--;
+      this._cacheKeys[index] = this._cacheKeys[this._cacheLength];
+    }
+  }
+  clear() {
+    this._cache = {};
+    this._cacheKeys = [];
+    this._cacheLength = 0;
```

Comment:
Any reason to not directly use `this. _cacheKeys.length`?

Reply:
Yeah iirc the idea here is.. you can't know the length if the array shrinks from deletes.

```js
> x = [1,2,3,4,5]
[ 1, 2, 3, 4, 5 ]
> x.length
5
> delete x[3]
true
> x
[ 1, 2, 3, <1 empty item>, 5 ]
> x.length
5
```